### ISSUE TITLE
Mysql `upsert()` fails with "Attempted to serialize empty result."

### ISSUE THREAD

**Issue by pslii** [prisma/prisma#18285]
Hi Prisma team, 

We've been occasionally seeing the following panic when we run upserts:
```
thread 'tokio-runtime-worker' panicked at 'Internal error: Attempted to serialize empty result.'
```
It only occurs when the row already exists in the database (i.e. we're doing an update) and in our testing, seems to only affect models that have multiple relations to other models. 

We started seeing this after upgrading to 4.7, but have repro'd the issue with versions up to 4.11; downgrading to 4.5 resolves the issue.

We've setup a up a repro of this issue here:
https://github.com/askchapter/prisma-upsert-debugging/tree/main/services/prisma-test (see README for instructions to run).

## Versions

| Name            | Version            |
|-----------------|--------------------|
| Node            | v16.13.0           | 
| OS              | darwin-arm64       |
| Prisma Client   | 4.11.0             |
| Query Engine    | 8fde8fef4033376662cad983758335009d522acb|
| Database        | mysql              |

## Logs
```
prisma:tryLoadEnv Environment variables loaded from /Volumes/git/prisma-upsert-debugging/services/prisma-test/.env
prisma:tryLoadEnv Environment variables loaded from /Volumes/git/prisma-upsert-debugging/services/prisma-test/.env
prisma:client dirname /Volumes/git/prisma-upsert-debugging/services/prisma-test/lib/__generated__/client
prisma:client relativePath ../../../prisma
prisma:client cwd /Volumes/git/prisma-upsert-debugging/services/prisma-test/prisma
prisma:client protocol graphql
prisma:client clientVersion 4.11.0
prisma:client clientEngineType library
prisma:client:libraryEngine internalSetup
prisma:client:libraryEngine:loader Searching for Query Engine Library in /Volumes/git/prisma-upsert-debugging/services/prisma-test/lib/.prisma/client
prisma:client:libraryEngine:loader Searching for Query Engine Library in /Volumes/git/prisma-upsert-debugging/services/prisma-test/src/__generated__/client
prisma:client:libraryEngine:loader loadEngine using /Volumes/git/prisma-upsert-debugging/services/prisma-test/src/__generated__/client/libquery_engine-darwin-arm64.dylib.node
prisma:client:libraryEngine library starting
prisma:client:libraryEngine library started
prisma:client:libraryEngine sending request, this.libraryStarted: true
```

Console logs with RUST_BACKTRACE=full
```
thread 'tokio-runtime-worker' panicked at 'Internal error: Attempted to serialize empty result.', query-engine/core/src/response_ir/ir_serializer.rs:54:40
stack backtrace:
   0:        0x11083e4f8 - _napi_register_module_v1
   1:        0x110204604 - <unknown>
   2:        0x11081c604 - _napi_register_module_v1
   3:        0x110841944 - _napi_register_module_v1
   4:        0x1108415a4 - _napi_register_module_v1
   5:        0x11084247c - _napi_register_module_v1
   6:        0x110841fe4 - _napi_register_module_v1
   7:        0x110841f78 - _napi_register_module_v1
   8:        0x110841f54 - _napi_register_module_v1
   9:        0x110ac298c - _napi_register_module_v1
  10:        0x11066d96c - _napi_register_module_v1
  11:        0x110131f50 - <unknown>
  12:        0x110132790 - <unknown>
  13:        0x110133428 - <unknown>
  14:        0x11014b9b4 - <unknown>
  15:        0x110148aec - <unknown>
  16:        0x11013a500 - <unknown>
  17:        0x110189334 - <unknown>
  18:        0x110163d14 - <unknown>
  19:        0x110173ee0 - <unknown>
  20:        0x11085999c - _napi_register_module_v1
  21:        0x11086509c - _napi_register_module_v1
  22:        0x11085898c - _napi_register_module_v1
  23:        0x1108583b8 - _napi_register_module_v1
  24:        0x11085e6c0 - _napi_register_module_v1
  25:        0x1108438e0 - _napi_register_module_v1
  26:        0x193d8426c - __pthread_deallocate
Encountered error
Invalid `prisma.employee.upsert()` invocation in
/Volumes/git/prisma-upsert-debugging/services/prisma-test/lib/index.js:10:27

```

## Client Snippet
```ts
  const employeeCreate: Prisma.EmployeeCreateInput & Prisma.EmployeeUpdateInput = {
    emailAddress: "test_employee@test.com",
  }

  await prisma.employee.upsert({
    where: {
      emailAddress: "test_employee@test.com"
    },
    create: employeeCreate,
    update: employeeCreate
  });

  // error occurs during this second upsert
  await prisma.employee.upsert({
    where: {
      emailAddress: "test_employee@test.com"
    },
    create: employeeCreate,
    update: employeeCreate
  });
```

## Schema
```prisma
generator client {
  provider        = "prisma-client-js"
  output          = "../src/__generated__/client"
  previewFeatures = ["fullTextIndex", "fullTextSearch"]
  binaryTargets   = ["native", "linux-musl"]
}

datasource db {
  provider     = "mysql"
  url          = env("DATABASE_URL")
  relationMode = "prisma"
}

model Employee {
  id                  String        @id @default(uuid())
  createdAt           DateTime      @default(now())
  updatedAt           DateTime      @updatedAt
  emailAddress        String        @unique
  applicationsOwned   Application[] @relation("owningEmployee")
  applicationsWritten Application[] @relation("writingEmployee")
}

model Application {
  id              String    @id @default(uuid())
  ownerId         String
  writerId        String?
  owningEmployee  Employee  @relation("owningEmployee", fields: [ownerId], references: [id], onUpdate: NoAction)
  writingEmployee Employee? @relation("writingEmployee", fields: [writerId], references: [id], onUpdate: NoAction)

  @@index([ownerId])
}
```

## Prisma Engine Query
```
{"X":{}}
```

**Comment by pslii** [comment]
related: #17279, #17305

**Comment by pslii** [comment]
Will also add that the upserts start working again when we comment out the relations
```prisma
model Employee {
  id                  String        @id @default(uuid())
  createdAt           DateTime      @default(now())
  updatedAt           DateTime      @updatedAt
  emailAddress        String        @unique
  // applicationsOwned   Application[] @relation("owningEmployee")
  // applicationsWritten Application[] @relation("writingEmployee")
}

model Application {
  id              String    @id @default(uuid())
  ownerId         String
  writerId        String?
  // owningEmployee  Employee  @relation("owningEmployee", fields: [ownerId], references: [id], onUpdate: NoAction)
  // writingEmployee Employee? @relation("writingEmployee", fields: [writerId], references: [id], onUpdate: NoAction)

  @@index([ownerId])
}
```

**Comment by pauledrupt** [comment]
same here

**Comment by pauledrupt** [comment]
only happens when you have two or more self-referencing relations

**Comment by janpio** [comment]
@pauledrupt Are you also using `relationMode=prisma` as P shared in his snippet?

**Comment by laplab** [comment]
Thank you for reporting the issue, @pslii! We are currently testing the fix for this problem. Would it be possible for you to check your code against this version of Prisma?
```
npm i prisma@5.13.0-dev.30
```
Bear in mind that this is a development version and should not be used in production. This is only to check if the issue is fixed.

**Comment by pslii** [comment]
@laplab confirmed that I'm not able to reproduce this anymore after updating to `5.13.0-dev.30`!

**Comment by laplab** [comment]
This issue was fixed in [Prisma 5.13.0](https://github.com/prisma/prisma/releases/tag/5.13.0).

**Comment by pauledrupt** [comment]
> @pauledrupt Are you also using `relationMode=prisma` as P shared in his snippet?

sorry, I didn't see this and yes
